api/src/main/java/io/grpc/TlsServerCredentials.java

```diff
@@ -196,36 +291,130 @@ public Builder keyManager(File certChain, File privateKey, String privateKeyPass
+     * just be a single entry and it implements {@link javax.net.ssl.X509KeyManager}.
+     */
+    public Builder keyManager(KeyManager... keyManagers) {
+      List<KeyManager> keyManagerList = Collections.unmodifiableList(new ArrayList<>(
+          Arrays.asList(keyManagers)));
+      clearKeyManagers();
+      this.keyManagers = keyManagerList;
+      return this;
+    }
+
+    private void clearKeyManagers() {
+      this.certificateChain = null;
+      this.privateKey = null;
+      this.privateKeyPassword = null;
+      this.keyManagers = null;
+    }
+
+    /**
+     * Indicates whether the server should expect a client's identity. Must not be {@code null}.
+     * Defaults to {@link ClientAuth#NONE}.
+     */
+    public Builder clientAuth(ClientAuth clientAuth) {
+      Preconditions.checkNotNull(clientAuth, "clientAuth");
+      this.clientAuth = clientAuth;
+      return this;
+    }
+
+    /**
+     * Use the provided root certificates to verify the client's identity instead of the system's
+     * default. Generally they should be PEM-encoded with all the certificates concatenated together
+     * (file header has "BEGIN CERTIFICATE", and would occur once per certificate).
+     */
+    public Builder trustManager(File rootCerts) throws IOException {
+      InputStream rootCertsIs = new FileInputStream(rootCerts);
+      try {
+        return trustManager(rootCertsIs);
+      } finally {
+        rootCertsIs.close();
+      }
+    }
+
+    /**
+     * Use the provided root certificates to verify the client's identity instead of the system's
+     * default. Generally they should be PEM-encoded with all the certificates concatenated together
+     * (file header has "BEGIN CERTIFICATE", and would occur once per certificate).
+     */
+    public Builder trustManager(InputStream rootCerts) throws IOException {
+      byte[] rootCertsBytes = ByteStreams.toByteArray(rootCerts);
+      clearTrustManagers();
+      this.rootCertificates = rootCertsBytes;
+      return this;
+    }
+
+    /**
+     * Have the provided trust manager verify the client's identity instead of the system's default.
+     * Although multiple are allowed, only the first instance implementing a particular interface is
+     * used. So generally there will just be a single entry and it implements {@link
+     * javax.net.ssl.X509TrustManager}.
+     */
+    public Builder trustManager(TrustManager... trustManagers) {
+      List<TrustManager> trustManagerList = Collections.unmodifiableList(new ArrayList<>(
+          Arrays.asList(trustManagers)));
+      clearTrustManagers();
+      this.trustManagers = trustManagerList;
+      return this;
+    }
+
+    private void clearTrustManagers() {
+      this.rootCertificates = null;
+      this.trustManagers = null;
+    }
+
     /** Construct the credentials. */
     public ServerCredentials build() {
-      if (certificateChain == null) {
+      if (certificateChain == null && keyManagers == null) {
```

Comment:
Should you also check that if ClientAuth is REQUIRE then the trustManager cannot be null?